test/lib/worker.test.js

```diff
@@ -405,13 +434,260 @@ describe('Worker library', () => {
           exchangeName,
           queueName
         }, {
+          channelCloseTimeout: 1,
           processExitTimeout: 1,
           logger
         });
       yield worker.listen();
       yield worker.close();
-      yield cb => setTimeout(cb, 500);
-      expect(process.exit.called).to.be.true();
+      expect(exitStub.args).to.deep.equal([
+        [0]
+      ]);
+    });
+
+    it('should not exit process on worker close (forceExit=false)', function* test() {
+      const worker = createWorkers([{
+        handle: _.identity,
+        validate: _.identity,
+        routingKey
+      }],
+        {
+          workerName,
+          amqpUrl,
+          exchangeName,
+          queueName
+        }, {
+          channelCloseTimeout: 1,
+          processExitTimeout: 1,
+          logger
+        });
+      yield worker.listen();
+      yield worker.close(false);
+      expect(exitStub.args).to.deep.equal([]);
+    });
+
+    it('should be idempotent (can be safely called twice)', function* test() {
+      const worker = createWorkers([{
+        handle: _.identity,
+        validate: _.identity,
+        routingKey
+      }],
+        {
+          workerName,
+          amqpUrl,
+          exchangeName,
+          queueName
+        }, {
+          channelCloseTimeout: 1,
+          processExitTimeout: 1,
+          logger
+        });
+      yield worker.listen();
+      yield worker.close();
+      yield worker.close();
+      expect(exitStub.args).to.deep.equal([
+        [0]
+      ]);
+    });
+
+    it('should finish handling current messages before closing a channel', function* test() {
+      const handlerStub = sandbox.stub()
+        .returns(() => new Promise(resolve => setTimeout(resolve, 20)));
+      const worker = createWorkers([{
+        handle: handlerStub,
+        validate: _.identity,
+        routingKey
+      }],
+        {
+          workerName,
+          amqpUrl,
+          exchangeName,
+          queueName
+        }, {
+          channelCloseTimeout: 50,
+          processExitTimeout: 1,
+          logger
+        });
+      yield worker.listen();
+      const message = new Buffer(JSON.stringify({ hello: 'world' }));
+      channel.publish(exchangeName, routingKey, message);
+      yield cb => setTimeout(cb, 10);
+      yield worker.close(false);
+
+      const remainingMessage = yield channel.get(formattedQueueName);
+      expect(remainingMessage).to.equal(false);
+
+      expect(handlerStub.args.map(arg => arg[0])).to.deep.equal([
+        { hello: 'world' }
+      ]);
+    });
+
+    it('should nack unfinished messages if the close timeout is over', function* test() {
+      const handlerStub = sandbox.stub()
+        .returns(() => new Promise(resolve => setTimeout(resolve, 20)));
```

Comment:
Maybe use a really bigger timeout to be really explicit about the test case?

Reply:
to have more sense it has to finish during the test but after the channel is close